# FIT5196 Task 1 in Assessment 2
#### Student Name: Zhiqing Shu
#### Student ID: 28217551

Date: 13/05/2018

Version: 1.0

Environment: Python 3.6.5 and Jupyter notebook

Libraries used:
* pandas (for pulling data out of HTML and XML files, included in Anaconda Python 3.6)
* numpy (for numpy array, included in Anaconda Python 3.6)
* re (for regular expression, included in Anaconda Python 3.6) 
* datetime (for manipulating dates and times，included in Anaconda Python 3.6) 
* sys (for system information, included in Anaconda Python 3.6)
* googlemaps (for getting location information, need to install before using!)
* requests (for HTTP request, included in Anaconda Python 3.6)

## Introduction
In this task, you are required to inspect and audit the data (dataset1_with_error.csv) to identify the data problems, and then fix the problems. Different generic and major data problems could be found in the data might include:
* Lexical errors
* Irregularities
* Violations of the Integrity constraint.
* Inconsistency

In the end, save the error-free dataset in dataset1_solution.csv. The number of records in your solution should be the same as the number of those in the input file.

### 1. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import re
import datetime
import sys
try:
    import googlemaps
except:
    import pip
    !pip install googlemaps
    import googlemaps
import requests

In [2]:
sys.version

'3.6.5 |Anaconda, Inc.| (default, Apr 26 2018, 08:42:37) \n[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]'

### 2. Load the data and store it in a Panda DataFrame

Since the data we need to load is a CSV file, so So we  use the <a href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html">read_csv()</a> function to load dataset1.


In [3]:
dataset1 = pd.read_csv("dataset1_with_error.csv")

### 3. Auditing and Cleansing the Job dataset

Starting by looking at the dimensionality of the data and a few lines of the data.

In [4]:
dataset1.head(5)

Id                                              Title   Location  \
0  12612628                        Engineering Systems Analyst    Dorking   
1  12612830                            Stress Engineer Glasgow    Glasgow   
2  12612844                   Modelling and simulation analyst  Hampshire   
3  12613049  Engineering Systems Analyst / Mathematical Mod...     Surrey   
4  12613647         Pioneer, Miser Engineering Systems Analyst     Surrey   

    ContractType ContractTime                       Company          Category  \
0  not available    permanent  Gregory Martin International  Engineering Jobs   
1  not available    permanent  Gregory Martin International  Engineering Jobs   
2  not available    permanent  Gregory Martin International  Engineering Jobs   
3  not available    permanent  Gregory Martin International  Engineering Jobs   
4  not available    permanent  Gregory Martin International  Engineering Jobs   

  Salary per annum        SourceName         OpenDate        CloseDate  
0            24996  cv-library.co.uk  20121103T000000  20121203T000000  
1            30000  cv-library.co.uk  20130108T150000  20130408T150000  
2            30000  cv-library.co.uk  20130726T150000  20130924T150000  
3            27504  cv-library.co.uk  20121214T000000  20130314T000000  
4            24996  cv-library.co.uk  20131025T000000  20131224T000000

In [5]:
dataset1.shape

(25077, 11)

In [6]:
dataset1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25077 entries, 0 to 25076
Data columns (total 11 columns):
Id                  25077 non-null int64
Title               25077 non-null object
Location            25077 non-null object
ContractType        25077 non-null object
ContractTime        25077 non-null object
Company             21242 non-null object
Category            25077 non-null object
Salary per annum    25077 non-null object
SourceName          25077 non-null object
OpenDate            25077 non-null object
CloseDate           25077 non-null object
dtypes: int64(1), object(10)
memory usage: 2.1+ MB


From above information, we can know that there are 25077 observations of 11 variables (but in Company, some observations are missing).

Use the `describe()` function to get the distribution of each variable (both numerical and categorial.

In [7]:
dataset1.describe()

Id
count  2.507700e+04
mean   6.664312e+07
std    5.195261e+06
min    1.261263e+07
25%    6.720830e+07
50%    6.836110e+07
75%    6.871371e+07
max    6.924767e+07

In [8]:
dataset1.describe(include='O')

Title Location   ContractType  \
count                                   25077    25077          25077   
unique                                  25077      482              3   
top     SQL Developer  SQL Server  Gloucester       UK  not available   
freq                                        1     3996          19499   

       ContractTime        Company Category Salary per annum     SourceName  \
count         25077          21242    25077            25077          25077   
unique            3           4879        8             1589             90   
top       permanent  UKStaffsearch  IT Jobs            35004  totaljobs.com   
freq          16194            248     7085             1011           5335   

               OpenDate        CloseDate  
count             25077            25077  
unique             2203             2400  
top     20130127T150000  20130526T120000  
freq                 27               23

Then we start to check based on the above information.

### 3.1 Check 'Id'

Using code below to check if all Id are in 8-digit.

In [9]:
#str(dataset1['Id'][1])

In [10]:
#Check if IDs are all 8-digit number
idList = []
for i in range(len(dataset1['Id'])):
    Id = re.findall(r'\d{8}', str(dataset1['Id'][i])[:9])
    idList.append(Id)
    
print(len(idList))

25077


And we can know there are 25077 unique Id by using `describe` function. So I believe there is no problem in 'Id'. 

### 3.2 Check 'Title'

Extract 'Title' data by `value_counts()` function.

In [11]:
dataset1.Title.value_counts()

SQL Developer  SQL Server  Gloucester                                    1
Teacher of Autism and Complex Needs  West London                         1
Associate Director Of Tax                                                1
GRADUATE SALES EXECUTIVE / NATIONAL ACCOUNT EXECUTIVE                    1
Assistant Capital Accountant                                             1
Full Training Given, Become a Hairdresser                                1
Java Developer (Scala / Haskell) Stockport, North West                   1
Creative Solutions Manager                                               1
Multiskilled Maintenance Engineer Leading manufacturing Group            1
Nursing Home Manager large home                                          1
Senior Exhaust Engineer                                                  1
Graduate Data / Pricing Analyst ****; Great Opportunity ****; Reading    1
PHP Developer  EYorks  ****  **** Per Day                                1
Senior SAS Analyst  Colle

There is no duplication in this column, and we can find there are some information hidden by `*`, by using commen sense, those `*` followed by `K` normally refer to the salary, but the others occurring alone are more complicated. I decide to keep these data unchanged since we do not have enough information to fill in.

### 3.3 Check 'Location'

In [12]:
dataset1.Location.value_counts()

UK                      3996
London                  2743
South East London       1433
The City                 558
Leeds                    343
Manchester               334
Surrey                   317
Central London           316
Reading                  300
Birmingham               292
West Midlands            273
East Sheen               222
Bristol                  222
Berkshire                202
Nottingham               196
Oxford                   190
Hampshire                184
Milton Keynes            180
Cambridge                172
Sheffield                160
Kent                     158
Newcastle Upon Tyne      156
Hertfordshire            155
Guildford                147
Aberdeen                 141
Leicester                139
Cheshire                 133
Oxfordshire              127
West Yorkshire           127
Belfast                  125
                        ... 
Linlithgow                 4
Reeding                    4
Oldbury                    4
Melksham      

Normally, 'Location' is likely to be misspelt, so we need to check if such problem exist.
There are lots of tools can be used to check if a location exist, considering the usability, I decided to use Google Maps API.

By calling `googlemaps.geocode(location,country)`, if this location exist in reality, we will get the detail information about it in JSON, while if the location not exist, Google Maps will match the most possible result, which can help us to detect misspelling location. The code and example as following:

In [13]:
#import googlemaps api, test the geocode usage
googlemaps = googlemaps.Client(key='AIzaSyBBxqEweHtJSYXm7OSIC9YwFoDFDnBcov4')
result = googlemaps.geocode('London,UK')
result #json

# address_components[] is an array containing the separate components applicable to this address.
# formatted_address is a string containing the human-readable address of this location.

[{'address_components': [{'long_name': 'London',
    'short_name': 'London',
    'types': ['locality', 'political']},
   {'long_name': 'London', 'short_name': 'London', 'types': ['postal_town']},
   {'long_name': 'Greater London',
    'short_name': 'Greater London',
    'types': ['administrative_area_level_2', 'political']},
   {'long_name': 'England',
    'short_name': 'England',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': 'United Kingdom',
    'short_name': 'GB',
    'types': ['country', 'political']}],
  'formatted_address': 'London, UK',
  'geometry': {'bounds': {'northeast': {'lat': 51.6723432, 'lng': 0.148271},
    'southwest': {'lat': 51.38494009999999, 'lng': -0.3514683}},
   'location': {'lat': 51.5073509, 'lng': -0.1277583},
   'location_type': 'APPROXIMATE',
   'viewport': {'northeast': {'lat': 51.6723432, 'lng': 0.148271},
    'southwest': {'lat': 51.38494009999999, 'lng': -0.3514683}}},
  'place_id': 'ChIJdd4hrwug2EcRmSrV3Vo6llI',
  'types'

I assume that misspelling location only account for a very small amount and will not occur frequently, so I choose those location occurring less than 10 times to check and creating a DataFrame to store these data

In [14]:
#Generate Location check dataframe
locationOccurrence = dataset1['Location'].value_counts().to_frame()
locationOccurrence.reset_index(inplace = True)
locationOccurrence.rename(columns={'Location': 'Occurrence','index': 'Location'}, inplace=True)
locationToCheck = locationOccurrence[locationOccurrence.Occurrence < 10]
locationToCheck.head()

Location  Occurrence
316      Accrington           9
317     South Brent           9
318        Bicester           9
319         Coscote           9
320  Bexhill-On-Sea           9

Running code below to get the misspelling location and the correct spelling.

In [16]:
# Using Geocode api to check the typos in Location
for location in locationToCheck.Location:
    location_pair = []
    result = googlemaps.geocode(location + ',UK')
    address = result[0]['formatted_address']
    if 'UK' in address and location.upper() not in address.upper():
        location_pair.append([location,address])
        print (location_pair)

[['Docklands', '4, Laybourne House, Admirals Way, Isle of Dogs, London E14 9UH, UK']]
[['St. Helens', 'St Helens, Saint Helens, UK']]
[['Leads', 'Leeds, UK']]
[['Surey', 'Surrey, UK']]
[['Nottinham', 'Nottingham, UK']]
[['Reeding', 'Reading, UK']]
[['Oxfords', 'Oxford, UK']]
[['St. Neots', 'St Neots, Saint Neots PE19, UK']]


We get 8 pairs. However, there is actually a place named `Docklands` in London (*London Docklands is the name for the riverfront and former docks in London. From Wikipedia*), so I think it is not misspelt. `St. Helens` and `St. Neots` are also correct. As a result, we just need to replace 5 misspellings.

In [17]:
dataset1.replace({'Location':{'Nottinham':'Nottingham','Leads':'Leeds','Surey':'Surrey',\
                             'Oxfords':'Oxford','Reeding':'Reading'}},inplace = True)

Running code above to correct all misspelling location name in dataset.

Check the result.

In [18]:
dataset1[dataset1.Location == 'Nottinham']
dataset1[dataset1.Location == 'Leads']
dataset1[dataset1.Location == 'Surey']
dataset1[dataset1.Location == 'Oxfords']
dataset1[dataset1.Location == 'Reeding']

Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []

### 3.4 Check 'ContractType' and 'ContractTime'

We have already known there are 3 unique value in 'ContractType' and 'ContractTime', then using `value_counts()` function to get these values and their distribution. 

In [19]:
dataset1.ContractType.value_counts()

not available    19499
full_time         4883
part_time          695
Name: ContractType, dtype: int64

In [20]:
dataset1.ContractTime.value_counts()

permanent        16194
not available     6212
contract          2671
Name: ContractTime, dtype: int64

The result looks good. However, in the dataset description carefully, the possible value should be **'non-specified'** rather than **'not available'**, so it will be better to make replacement by code below.

In [21]:
dataset1.ContractTime.replace({'not available': 'non-specified'},inplace=True)

In [22]:
dataset1.ContractType.replace({'not available': 'non-specified'},inplace=True)

In [23]:
print(dataset1.ContractType.value_counts())
print(dataset1.ContractTime.value_counts())

non-specified    19499
full_time         4883
part_time          695
Name: ContractType, dtype: int64
permanent        16194
non-specified     6212
contract          2671
Name: ContractTime, dtype: int64


### 3.5 Check 'Company'

By using `describe()` function, we have already known there are missing values in 'Company', but unfortunately we are not able to fill in them based on the imformation we have, so just let it go.

In [24]:
dataset1[dataset1.Company.isnull()]

Id                                              Title  \
51     34479775                                  Credit Controller   
55     34790490                                     Java Developer   
66     38209887                     Transaction Services Executive   
67     38209888                                    Audit Executive   
68     38209893                            Audit Assistant Manager   
70     38758190                                  Account Executive   
78     40880208                                 Compliance Manager   
81     41879934   Future Leaders Development Programme  Technology   
105    46626508                        Support Worker Job Chepstow   
106    46626509                     Support Worker Job Abergavenny   
107    46626511                          Staff Nurse Job Berkshire   
108    46626514                    Pharmacy Technician Job Norwich   
109    46626520                         Support Worker Job Cardiff   
110    46626537               RGN Registered Nurse Job Southampton   
111    46626546                      RGN Nurse Job Bury St Edmunds   
112    46626547                          Scrub Nurse Job Braintree   
113    46626549                Healthcare Assistant Job Nottingham   
114    46626555       Consultant Clinical Psychologist Job Norwich   
115    46626557  Operating Department Practitioner ODP Jobs Not...   
116    46626559                 RGN Prison Nurse Job Staffordshire   
117    46626566             Clinical Research Nurse Job Nottingham   
118    46626571                    RGN Registered Nurse Job Durham   
119    46626573     Prison Nurse Practitioner Jobs Nottinghamshire   
120    46626574    Care Assistants and Support Workers Job Glasgow   
121    46626799           Saturday only Night Nurse Job Winchester   
122    46626802                Nursing Home Staff Nurse Job Surrey   
123    46626826                        Substance Misuse Nurse Kent   
124    46626831        Registered Mental Health Nurse (RMN) Durham   
125    46626832    Registered Mental Health Nurse (RMN) Sunderland   
126    46626833          Registered General Nurse (RGN) Sunderland   
...         ...                                                ...   
23862  69166844                       Nurse : Surgical Staff Nurse   
23863  69166851           Nurse : Deputy Nursing Home Manager ****   
23864  69167060  Nurse : Registered General Nurse Jobs in Leigh...   
24582  69189730            Corporate Tax Compliance Senior Manager   
24586  69190364                        Inclusion Support Assistant   
24592  69190507      Junior Database Administrator  Holton St Mary   
24596  69190934                                Graphics Technician   
24629  69191688                  Accounts Assistant (MultiLingual)   
24631  69191782             Domestic Violence/Abuse Support Worker   
24633  69192290                        Progress Webspeed Developer   
24634  69192326  Property Sales Consultant / Senior Sales Negot...   
24635  69192551            FT Skilled Senior Carers  Days & Nights   
24653  69193530                                   **** x Caretaker   
24686  69196406  APPOINTMENT OF INDEPENDENT MEMBERS TO CARDIFF ...   
24723  69200075                         Trainee Marketing Advisors   
24752  69203584                                  Dividends Analyst   
24753  69203586                                Settlements Analyst   
24754  69203620                               Client Admin Support   
24825  69208639         Resourcing Coordinator (Full or Part Time)   
24921  69215963                  Solar Sales Consultant  Inverness   
24923  69216134                            Skilled Specialist Chef   
24924  69216174                 Project Manager (Supported Living)   
25008  69220336                           Trainer  NVQ Level 3 BIT   
25009  69220384           Appointment Maker  Temporary / Permanent   
25012  69220596               Promotions Sales Manager  Short Term   
25037  69227935  Commun

### 3.6 Check 'Category'

View the 'Category' data. 

In [25]:
dataset1.Category.value_counts()

IT Jobs                             7085
Healthcare & Nursing Jobs           4334
Engineering Jobs                    3458
Accounting & Finance Jobs           3099
Sales Jobs                          2609
Hospitality & Catering Jobs         2124
Teaching Jobs                       1378
PR, Advertising & Marketing Jobs     990
Name: Category, dtype: int64

It seems no obvious error existing in this column.

### 3.7 Check 'Salary per annum'

In [26]:
dataset1['Salary per annum'].value_counts()

35004    1011
30000     888
39996     849
45000     747
32496     744
37500     707
24996     690
50004     616
27504     582
42504     549
54996     448
24000     444
20004     399
22500     387
47496     364
60000     320
18000     315
21996     296
26004     286
21000     284
18996     283
27996     278
52500     275
27000     258
29004     256
23004     247
17004     244
28500     215
23496     207
26496     206
         ... 
38148       1
29364       1
9228        1
7512        1
34452       1
12636       1
29724       1
24216       1
56748       1
25584       1
31032       1
18264       1
12240       1
59256       1
55824       1
45024       1
9972        1
77880       1
24984       1
39432       1
15492       1
71K         1
6684        1
20328       1
39K         1
49836       1
60900       1
25008       1
23124       1
15852       1
Name: Salary per annum, Length: 1589, dtype: int64

Basically, `Salary per annum` are presented as 5 digits. While, some salary are recorded as '??K', here K presenting 1000, or recorded as a range in decimal. We need to unify the format of `Salary per annum`.

In [27]:
dataset1[dataset1['Salary per annum'].str.contains('K')]

Id                                              Title  \
133    46626860            Registered Home Manager Job Bournemouth   
238    46627928           Care Home Manager Job North London ****K   
305    46628805                     Home Care Workers Berkhampsted   
596    46634306                  Home Manager Mental Health 11 bed   
830    46637596                     Staff Nurse South Shields ****   
952    49065458              Service Manager Learning Disabilities   
996    49845058  Chef de Partie  Fresh Food Pub  Good Reputatio...   
1081   51285942                     Senior Social Worker Fostering   
1379   55408955                      Senior Applications Developer   
1410   55409587                          MVNO Enterprise Architect   
1621   56814478                   Chef de Partie  **** AA Rosettes   
2063   60685515          Staff Nurse Elderly Care Stockton on Tees   
3274   65317743  Assistant General Manager  New Opening  Aberde...   
4074   66425925              Head of Unit Dementia Buckinghamshire   
4144   66434456                         Commercial Project Manager   
4610   66601462                                  Nurse RGN Belfast   
4844   66674979         Chartered Mechanical / Structural Engineer   
5424   66923385  C++/C Developer  ****x 3 Months contracts  Wil...   
5446   66924597  Business Applications and Information Analyst ...   
5638   66930808      Graduate Trainee Media Sales Executive London   
5682   66934296                                     RMN  Doncaster   
5877   66980781                      Technical Development Manager   
6304   67210652  IT Support Engineer ndash; Windows, Lotus Note...   
6385   67244580                        Head of IT / Infrastructure   
6639   67323089    Senior Infrastructure Engineer  VMWARE  Bristol   
6983   67389525                                  Aircraft Mechanic   
7030   67445417  Technical Sales to Medical Sector  Northern En...   
7508   67748366              Deputy Nurse/ Clinical Lead  Carlisle   
7705   67770359                           IT Support Desk Operator   
8047   67894716              Business Development Manager  Defence   
...         ...                                                ...   
17712  68697764                        Shift Manager HV Authorised   
18121  68705063          senior chef de partie or junior sous chef   
18329  68707828  Graduate/ Trainee Estate Agent  South West London   
18550  68710330                           Electronic Test Engineer   
19231  68784342  Brand Manager / Product Manager  Health and Be...   
19312  68800514  Experienced Account Manager needed – Retail Cl...   
19588  68806012  Solutions Architect  Surrey  **** , ****  ****...   
19624  68806247  National Marketing Business Development Manage...   
19690  68806712                                  SQL Database Lead   
19808  68807719                                Mechanical Graduate   
20046  68833239                            Business Change Analyst   
20398  68844663                   Unique Test Lead position London   
20679  68995331                                   Claims Validator   
21322  69006838                                        Physics NQT   
21666  69012121                    Adult Speech Language Therapist   
21744  69018023  PHP WEB DEVELOPER / PHP PROGRAMMER / SOFTWARE ...   
21951  69022222  Software Test Engineer **** East Anglia  Embedded   
22029  69022816         PR and Events Manager (Maternity Contract)   
22420  69039153  Business Analyst (Debt Finance, Corporate Bank...   
22465  69039760  Project Manager (Software development) near Ma...   
22493  69039997        Graduate Software Developer – online gaming   
23478  69091714                                Web Support Analyst   
23984  69169854                   Marketing Manager  Food industry   
24027  69170445  Graduate Trainee Financial Risk Trader  ****k ...   
24170  69172786         Qualified Social Workers for ALL of London   
24364  69182561        

Replacing `'K'` with `000`.

In [28]:
# Replace K with 000
dataset1['Salary per annum'].replace('K','000',regex = True,inplace = True)

Check if there is any data presented in different way.

In [31]:
dataset1[dataset1['Salary per annum'].str.contains('\D')]
# 'Series' object has no attribute 'contains'

Id                                              Title  \
647    46634923        RGN Nurse Hull Days or Nights **** per hour   
896    48082563  Senior Chef de Partie  One AA Rosette Hotel  T...   
980    49689021                                 Assessment Officer   
1062   51061201              Community Home Workers Melton Mowbray   
1141   52489008                                      RGN Care Home   
1292   55400384                    Live in Care Worker Jobs Exeter   
1525   56209766                                 Nurse Advisor COPD   
1797   58533485          Senior Embedded Controls Systems Engineer   
1822   58759123  Helpdesk / Desktop Support analystExcel, Word,...   
2051   60684270              Door to door / Field Sales Canvassers   
2118   61445890                              Mental Health Advisor   
2602   64145428                             DEPUTY MANAGER READING   
2724   64573377                 Deputy Care Manager / Care Manager   
2889   64809245  Contract Mechanical Draughtsperson (Huddersfield)   
2923   64813318  Manufacturing Engineer (Mechanical / Pumps / CNC)   
3057   65171402                                 Complaints Manager   
3186   65188754       Technical Sales Engineer (ElectroMechanical)   
3587   65879969                               Contract Salesperson   
3927   66197678  Front End Web Developer – (HTML, CSS, Javascript)   
3973   66367073                      Financial Controller (Assets)   
4705   66608673  INFRASTRUCTURE SUPPORT – OPERATIONS – CLOUD, S...   
5708   66935869         Account Manager, Account Executive  Energy   
5988   67017625                      Contracts Manager (Permanent)   
6225   67153390   Electrical Field Service Engineer x**** (Cranes)   
6270   67208324            Building Services Supervisor in Chelsea   
6710   67333923                               HVAC Design Engineer   
6977   67389276                     Lead Business Analyst, Banking   
7141   67638912  IT Sales Executive / New Business Development ...   
7296   67652461  MS SharePoint Senior Developer, R****k, Jo bur...   
7347   67652829               IT Equipment Installations Assistant   
...         ...                                                ...   
17906  68702082  Electronic / Electrical Test Engineer (Lithium...   
17983  68703232                             Commercial Team Leader   
18002  68703556             Service Desk / Desktop Support Analyst   
18174  68706043  Manager/ Unit Manager/ Store Manager/ Catering...   
18905  68715405  TECHNICAL LEAD CASPNET MVC SCRUM TDD HOSPITALI...   
18964  68716258       Project Manager  Southampton  ****K to ****K   
19270  68787221  IT Sales Executive  Accrington  ****k to ****k...   
19445  68804934                   Mechanical Estimator  Birmingham   
19548  68805804          Relocate to the US? QA Software Engineers   
20070  68834330                Field Service Connectivity Engineer   
20175  68837150                Restaurant Manager  ****AA Rosettes   
20854  68999922          INTERNAL SALES  **** Year Maternity Cover   
21064  69004198                           Helpdesk Apprenticeships   
21242  69005600            Outsource Manager / Employer Engagement   
21247  69005706  FRENCH AND SPANISH TEACHER  QTS  **** **** PER...   
21313  69006748                           Part Time Year 6 Teacher   
21734  69017589  BI Manager, SQL Server Financial ServicesHertf...   
22283  69036978  Application Support Analyst (legal sector)  Lo...   
22418  69039080          Oracle & SQL DBA  Investment Bank  London   
23111  69080469                                  Catering Engineer   
23152  69080768              Drilling Equipment Service Technician   
23207  69081257                    Principal/Senior ITS Consultant   
23365  69083476                                  Auto CAD Designer   
23557  69118140  Sales Manager Drain Cleaning & Inspection Equi...   
23619  69122549                                   Credit Assistant   
23815  69162548        

Since the majority of salary are wrote in integer, so we need to modify unformated salary into integer. At the same time, we can check if the range is represented as `lower - upper`.

In [39]:
# remove extra decimal place check the range validity 
for index,data in enumerate(dataset1['Salary per annum']):
    if '-' in dataset1.loc[index]['Salary per annum']:
        salary = dataset1.loc[index]['Salary per annum'].split('-')
        lower = round(float(salary[0]))
        upper = round(float(salary[1]))
        if upper < lower:
            print (dataset1.loc[i])
        else:
            dataset1.loc[index,'Salary per annum'] = str(lower) + '-' + str(upper)

No mis-ranged data and the format has been modified.

In [40]:
# Randomly choose one row to check result
dataset1.loc[2118]['Salary per annum']

'24647-27241'

In [43]:
type(dataset1['Salary per annum'].loc[2118])

str

### 3.9 Check 'SourceName'

`SourceName` is *The website where the job position is advertised*, so we need to check if all websites exist.

We choose the website occurring less than 10 times as suspicious source name.

Creating a DataFrame to store these suspicious source name.

In [44]:
#Generate SourceName check dataframe
sourceOccurrence = dataset1['SourceName'].value_counts().to_frame()
sourceOccurrence.reset_index(inplace = True)
sourceOccurrence.rename(columns={'SourceName': 'Occurrence','index': 'SourceName'}, inplace=True)
sourceToCheck = sourceOccurrence[sourceOccurrence.Occurrence < 10]
sourceToCheck.head()

SourceName  Occurrence
75      legalprospects.com           8
76   renewablescareers.com           8
77           wikijob.co.uk           5
78               3desk.com           4
79  jobsinsocialwork.co.uk           4

We use the reponse status code to make sure if the website exist. *In computer network communications, `404` is used to indicate that the client was able to communicate with a given server, but the server could not find what was requested.*, so if we get `404`, it means the website does not exist.

In [45]:
#404 means page not found(invalid page)
for link in sourceToCheck.SourceName:
    try: 
        r = requests.head("http://" + link)
        r.status_code != 404
    except :
        print(link)

jobcareer
monashstudent


Also, we need to check if there is any email address.

In [46]:
dataset1[dataset1['SourceName'].str.contains('@')]

Id                                             Title  \
15379  68672352  Digital Account Manager  **** plus a Great Bonus   

                Location   ContractType ContractTime      Company    Category  \
15379  South East London  non-specified    permanent  Blu Digital  Sales Jobs   

      Salary per annum         SourceName         OpenDate        CloseDate  
15379            27504  admin@caterer.com  20120516T000000  20120715T000000

Replacing the invalid source name with correct website address.

In [47]:
dataset1.replace({'SourceName':{'monashstudent':'monash.edu/students','jobcareer':'jobcareer.com',
                                'admin@caterer.com' : 'caterer.com'}},inplace = True)

Checking if the replacement is executed successfully.

In [48]:
dataset1[dataset1.SourceName == 'monashstudent']

Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []

In [49]:
dataset1[dataset1.SourceName == 'jobcareer']

Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []

In [50]:
dataset1[dataset1.SourceName == 'admin@caterer.com']

Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []

### 3.10 Check 'OpenDate' and 'CloseDate'

According to the data description: 

*The opening time for applying for the advertised job position, e.g., 20120104T150000, means 3pm, 4th January 2012.*

so we can use `regular expression` to extract date and time. 

A normal issue occurs in date is invalid date, for example, month overs 12 or day overs 31. So we need to check this kind of problem first.

In [51]:
#Checking Date validity
for date in dataset1.OpenDate:
    if int(re.findall('\d{4}(\d{2}).*', date)[0]) > 12:
        print(date)

20131803T120000
20132606T000000
20122003T150000
20121512T150000
20133004T150000
20131908T000000
20121406T120000
20131509T000000
20132901T150000
20132108T120000


In [52]:
for date in dataset1.OpenDate:
    if int(re.findall('\d{4}\d{2}(\d{2})', date)[0]) > 31:
        print(date)

In [53]:
for date in dataset1.CloseDate:
    if int(re.findall('\d{4}(\d{2}).*', date)[0]) > 12:
        print(date)

In [54]:
for date in dataset1.CloseDate:
    if int(re.findall('\d{4}\d{2}(\d{2})', date)[0]) > 31:
        print(date)

By running above code, we found there are some invalid data in `OpenDate` and by observing these data, we can know `Month` and `Day` is mislocated, and we need to exchange them.

In [62]:
# create a function to get Year, Month, Day and Time, and change the location of Month and Day
def modifyDate(date):
    mDate = re.findall('(\d{4})(\d{2})(\d{2})(T\d{2}\d{2}\d{2})',date)
    return mDate[0] + mDate[2] + mDate[1] + mDate[3]

In [63]:
# replace the wrong date with the correct one
for index,date in enumerate(dataset1.OpenDate):
    if int(re.findall('\d{4}(\d{2}).*', date)[0]) > 12:
        dataset1.loc[index,'OpenDate'] = modifyDate(date)

Apart from above issue, another problem may happen is that `OpenDate` is later then `CloseDate`.

Creating a function `checkDateValid(dataset)` to check if this problem exist.

In [64]:
# Check Date range validity
def checkDateValid(dataset):
    invalid = []
    for index in range(len(dataset)):
        closeDate = pd.to_datetime(dataset.loc[index,'CloseDate'],format="%Y%m%dT%H%M%S")
        openDate = pd.to_datetime(dataset.loc[index,'OpenDate'],format="%Y%m%dT%H%M%S")
        if closeDate < openDate:
            invalid.append([index,openDate - closeDate])
    return invalid

In [65]:
checkDateValid(dataset1)

[[6659, Timedelta('90 days 00:00:00')],
 [9568, Timedelta('30 days 00:00:00')],
 [11043, Timedelta('30 days 00:00:00')],
 [12473, Timedelta('60 days 00:00:00')],
 [19142, Timedelta('90 days 00:00:00')],
 [24206, Timedelta('90 days 00:00:00')],
 [24297, Timedelta('90 days 00:00:00')],
 [25039, Timedelta('30 days 00:00:00')]]

There are abnormal date, and the interval is quite interesting, 30, 60, 90.

By running following code, we find the majority interval between `OpenDate` and `CloseDate` is also 30, 60 and 90, so I think the interval of problem data is valid, we just need to exchange `OpenDate` and `CloseDate`.

In [66]:
dataset1['Period'] = (pd.to_datetime(dataset1.CloseDate,format="%Y%m%dT%H%M%S")\
                      - pd.to_datetime(dataset1.OpenDate,format="%Y%m%dT%H%M%S"))

In [67]:
dataset1.head()

Id                                              Title   Location  \
0  12612628                        Engineering Systems Analyst    Dorking   
1  12612830                            Stress Engineer Glasgow    Glasgow   
2  12612844                   Modelling and simulation analyst  Hampshire   
3  12613049  Engineering Systems Analyst / Mathematical Mod...     Surrey   
4  12613647         Pioneer, Miser Engineering Systems Analyst     Surrey   

    ContractType ContractTime                       Company          Category  \
0  non-specified    permanent  Gregory Martin International  Engineering Jobs   
1  non-specified    permanent  Gregory Martin International  Engineering Jobs   
2  non-specified    permanent  Gregory Martin International  Engineering Jobs   
3  non-specified    permanent  Gregory Martin International  Engineering Jobs   
4  non-specified    permanent  Gregory Martin International  Engineering Jobs   

  Salary per annum        SourceName         OpenDate        CloseDate  Period  
0            24996  cv-library.co.uk  20121103T000000  20121203T000000 30 days  
1            30000  cv-library.co.uk  20130108T150000  20130408T150000 90 days  
2            30000  cv-library.co.uk  20130726T150000  20130924T150000 60 days  
3            27504  cv-library.co.uk  20121214T000000  20130314T000000 90 days  
4            24996  cv-library.co.uk  20131025T000000  20131224T000000 60 days

In [68]:
dataset1.Period.describe()

count                      25077
mean     53 days 04:26:50.718985
std      25 days 19:15:46.166444
min           -90 days +00:00:00
25%             30 days 00:00:00
50%             60 days 00:00:00
75%             60 days 00:00:00
max             90 days 00:00:00
Name: Period, dtype: object

Running following code to exchange `OpenDate` and `CloseDate`.

In [69]:
for index in range(len(dataset1)):
    closeDate = pd.to_datetime(dataset1.loc[index,'CloseDate'],format="%Y%m%dT%H%M%S")
    openDate = pd.to_datetime(dataset1.loc[index,'OpenDate'],format="%Y%m%dT%H%M%S")
    if closeDate < openDate:
        originalOpenDate = dataset1.loc[index]['OpenDate']
        originalCloseDate = dataset1.loc[index]['CloseDate']
        dataset1.loc[index,'OpenDate'] = originalCloseDate
        dataset1.loc[index,'CloseDate'] = originalOpenDate

Checking if all issues are solved. 

In [70]:
checkDateValid(dataset1)

[]

### 4. Save to CSV file

Dropping column `Period`.

In [71]:
dataset1 = dataset1.drop(columns=['Period'])

Check result.

In [72]:
dataset1.head()

Id                                              Title   Location  \
0  12612628                        Engineering Systems Analyst    Dorking   
1  12612830                            Stress Engineer Glasgow    Glasgow   
2  12612844                   Modelling and simulation analyst  Hampshire   
3  12613049  Engineering Systems Analyst / Mathematical Mod...     Surrey   
4  12613647         Pioneer, Miser Engineering Systems Analyst     Surrey   

    ContractType ContractTime                       Company          Category  \
0  non-specified    permanent  Gregory Martin International  Engineering Jobs   
1  non-specified    permanent  Gregory Martin International  Engineering Jobs   
2  non-specified    permanent  Gregory Martin International  Engineering Jobs   
3  non-specified    permanent  Gregory Martin International  Engineering Jobs   
4  non-specified    permanent  Gregory Martin International  Engineering Jobs   

  Salary per annum        SourceName         OpenDate        CloseDate  
0            24996  cv-library.co.uk  20121103T000000  20121203T000000  
1            30000  cv-library.co.uk  20130108T150000  20130408T150000  
2            30000  cv-library.co.uk  20130726T150000  20130924T150000  
3            27504  cv-library.co.uk  20121214T000000  20130314T000000  
4            24996  cv-library.co.uk  20131025T000000  20131224T000000

Save to CSV file and name it `dataset1_solution.csv`.

In [73]:
dataset1.to_csv('dataset1_solution.csv',encoding='utf-8',index=False)

### 5. Reference

- pandas. Retrieved from: http://pandas.pydata.org/pandas-docs/stable/
- Google map apis. Retrieved from: https://developers.google.com/maps/documentation/embed/
- python request.  Retrieved from: http://docs.python-requests.org/en/master/
- python datetime Retrieved from:https://docs.python.org/3/library/datetime.html